Issue #4 - Add initial content structure;

firasrg opened the pull request:
title: initial improvements on 1st parts;

comments:
- prelude.md & getting_started.md: expand intro by adding old-school developers and professionals;
- benefits_of_java.md: new file to talk about major benefits in Java;

firasrg committed:
Issue #4 - Add initial content structure;

title: initial improvements on 1st parts;

comments:
- prelude.md & getting_started.md: expand intro by adding old-school developers and professionals;
- benefits_of_java.md: new file to talk about major benefits in Java;

firasrg commented:
Hello reviewers! This is an initial PR, I want to show you what I've done so far and discuss, before I go further. Thanks

firasrg commented on src/SUMMARY.md:
```diff
@@ -35,6 +35,8 @@
     - [Inferred Types](./variables/inferred_types.md)
     - [Challenges](./variables/challenges.md)
 
+- [Benefits of java](./benefits_of_java.md)
```
FYI, I think that showing the benefits of Java is a good idea, especially at the present time as some new technologies have become increasingly dominant in the ideas of developers and organizations, so the importance of Java has become invisible in the required way, so the public and the private should be reminded of its importance and benefits. 

Also, to further encourage the reader and prepare him to continue reading and liking the book

bowbahdoe replied:
instead of "Benefits of Java" - "Why Java?" would be more apt.

firasrg commented on start.sh:
```diff
@@ -0,0 +1 @@
+mdbook serve
```
This is to remind new contributors about the required command to deploy the website locally. This is an initiative, we could add more commands or change this one as we go. 💡

bowbahdoe replied:
1. This only works on mac and linux
2. How is this one command any easier or harder to remember than `mdbook serve`?
3. Do you anticipate any more commands being added?

firasrg replied:
> This only works on mac and linux

it works on Windows Powersheel too, I've tried it. 

--- 
> How is this one command any easier or harder to remember than mdbook serve?

As I said, its just an initiative to store commands in files, and thats the 1st one.

---
> Do you anticipate any more commands being added?
yeah why not?

Taz03 replied:
if theres only one command i'd disagree to make a start.sh for it

bowbahdoe commented on src/benefits_of_java.md:
```diff
@@ -0,0 +1,29 @@
+## Object Oriented
+Java is a popular programming language that is based on [the object-oriented programming (OOP)](https://www.techtarget.com/searchapparchitecture/definition/object-oriented-programming-OOP) paradigm. This means that Java code is defined in classes, which are like blueprints that describe the properties and behaviors of objects.
+
+Compared to [procedural programming languages](https://www.geeksforgeeks.org/differences-between-procedural-and-object-oriented-programming) that use routines or methods without classes, Java's use of OOP allows for better code organization and reuse. However, Java also allows for [functional programming](#functional) within a class, which can make certain operations more concise and readable.
```
The target audience doesn't have context on what a procedural language is

firasrg replied:
Yeah, I assume that's true, that's why I've added the link there ⛓

bowbahdoe commented on src/prelude.md:
```diff
@@ -9,12 +9,12 @@ got away without having to.
+At the end of each section, you will find challenges that are designed to help reinforce your understanding of the concepts you saw. We recommend you to attempt these challenges before moving to next section(s).
 
-I see you. You are not stupid. This is not your fault, it is theirs.[^probably]
+We wish you the best of luck, and we think you're strong enough to get better as you go.
```
I do agree this part could use a more positive tone. This is a big vague and hunger-gamesy though

firasrg replied:
I think, that what I put here is better than saying "You are not stupid ..."

bowbahdoe commented on src/getting_started.md:
```diff
@@ -2,14 +2,13 @@
 
-If you are in school and your teacher has helped you get set up in some other
-way it is okay to skip this section and just do it the way you were shown.
+If you are currently learning Java in school and your teacher has already given you a different way to set up, it is ok to skip this section and just follow the way you were taught. Similarly, if you are a professional who already has experience setting up a development environment, you may choose to skip this section.
```
Professionals know they can skip set up. Remember, write to one audience.

firasrg replied:
As Professional the previous expression can make me feel that the book is intended for beginners as **primary** audience. I may exit the whole book directly right after reading the _getting started_.

bowbahdoe commented on src/getting_started.md:
```diff
@@ -2,14 +2,13 @@
 
-For the purposes of this book, I am going to reccomend that you
+For the purposes of this book, I am going to recommend that you
 start with [replit.com](https://replit.com).
 
-It requires an internet connection and you will have to make an account, but
+It requires an internet connection, and you will have to make an account, but
```
Ok yeah grammar fixes

bowbahdoe commented on src/benefits_of_java.md:
```diff
@@ -0,0 +1,29 @@
+
+Java has some key benefits that you’ll need to know 😉.
+
+## Object Oriented
```
No to this entire section. 

These arguments are half marketing fluff, half stuff that a beginner wouldn't have the context to understand. I don't think this is a problem with your writing style, but just in the content you want to convey.

If you want to have a "why java" section, focus on stuff people might actually care about.

- Tons of software written in it, it is part of the foundations of the world
- Job opportunities
- Minecraft mods
- "you have to start somewhere"
-  etc.

bowbahdoe commented on src/benefits_of_java.md:
```diff
@@ -0,0 +1,29 @@
+# Benefits of Java
+
+Java has some key benefits that you’ll need to know 😉.
```
Don't use emojis like this unless you are texting someone you want to have sex with.

bowbahdoe commented on src/benefits_of_java.md:
```diff
@@ -0,0 +1,29 @@
+Java is a compiled language that generates bytecode, which can be interpreted by the [Java Virtual Machine (JVM)](https://www.javatpoint.com/jvm-java-virtual-machine) on any platform . One of the benefits of this approach is that Java code only needs to be compiled once, and it can then run on any operating system that has a JVM installed. This concept is commonly referred to as ["write once, run ...
+
+## Strong
+Java has a significant advantage over some popular languages such as [C++](https://www.techtarget.com/searchdatamanagement/definition/C) in terms of memory management. Unlike C++, Java manages memory on its own and automatically performs [garbage collection](). This feature makes Java more robust and helps prevent memory leaks that can lead to errors and program crashes in C++.
```
Would you actually have enough context to understand a c++ comparison if you didn't come out of a time machine from the 1990s

firasrg replied:
Well, in my old study days we've been comparing Java most of the time to C++ or C# thats why.

bowbahdoe commented on src/benefits_of_java.md:
```diff
@@ -0,0 +1,29 @@
+Java has a significant advantage over some popular languages such as [C++](https://www.techtarget.com/searchdatamanagement/definition/C) in terms of memory management. Unlike C++, Java manages memory on its own and automatically performs [garbage collection](). This feature makes Java more robust and helps prevent memory leaks that can lead to errors and program crashes in C++.
+
+## Secured
+Java code runs inside the JVM! This creates a sandbox that makes it harder for code to do malicious behaviors on the computer it's running on.
```
The what? Tons of languages have VMs

firasrg replied:
I didn't know that actually lol

bowbahdoe commented on src/benefits_of_java.md:
```diff
@@ -0,0 +1,29 @@
+
+Functional programming is a programming paradigm that emphasizes the use of functions as the primary building blocks of a software. In Java, one of the key features of functional programming is the use of [lambda expressions](todo), which are anonymous functions that can be passed around as arguments to other functions or stored in variables. (You'll see more about these in [Lambda's](todo) chapt ...
+
+## Independent
```
Tons of languages have this

firasrg replied:
certainly. I think it is very useful to show that Java also has this especially for beginners or those who don't know Java yet, or those who have misconceptions

bowbahdoe commented on src/benefits_of_java.md:
```diff
@@ -0,0 +1,29 @@
+### Encapsulation
+You can use access modifiers to control access to your class members (fields, methods, and nested classes). This is an important feature that allows you to protect your data from unintended access and modification.
+
+## Functional
```
They won't learn how or what this is until way later and they have no context on what non-functional code would be

firasrg replied:
Yeah, this is a general point, we can say "We'll discuss about these deeply in chapter X". 💡

bowbahdoe commented on src/benefits_of_java.md:
```diff
@@ -0,0 +1,29 @@
+
+As you learn Java, it is important to understand the principles of object-oriented programming and how to use classes and objects effectively. Additionally, gaining familiarity with functional programming concepts can help you write more concise and efficient code in certain situations.
+
+### Encapsulation
```
who cares? every language has mechanisms for encapsulation

firasrg replied:
We can show here how Java takes care about encapsulation using its keyword and class structure styles. Otherwise, remove this :D

bowbahdoe commented on src/benefits_of_java.md:
```diff
@@ -0,0 +1,29 @@
+
+Compared to [procedural programming languages](https://www.geeksforgeeks.org/differences-between-procedural-and-object-oriented-programming) that use routines or methods without classes, Java's use of OOP allows for better code organization and reuse. However, Java also allows for [functional programming](#functional) within a class, which can make certain operations more concise and readable.
+
+As you learn Java, it is important to understand the principles of object-oriented programming and how to use classes and objects effectively. Additionally, gaining familiarity with functional programming concepts can help you write more concise and efficient code in certain situations.
```
> it is important to understand the principles of object-oriented programming

It really isn't. Think about it. 

1. They don't know enough to actually know these "principles"
2. These "principles" are rules of thumb from the 90s and are super super nuanced. 
3. They don't have enough mechanical knowledge to understand anything of this
4. You can certainly learn Java without buying in to any particular design style.

firasrg replied:
We can add links, some details or diagrams to show what are these. Even if the principles are _rules of thumb_ from 90s, they must have evolutions! 

I understand this an early stage, but there are some ways to simplify. I can do more

bowbahdoe reviewed (changes requested):
I disagree with the material content in the "why java" section

firasrg committed:
initial improvements - Apply PR review

comments:
- rename section "benefits_of_java" to "Why Java?"
- Many improvements in  "Why Java?" page.
- start.sh removed;

firasrg commented:
Hello @bowbahdoe ! I've made a new update, please evaluate it and give me your feedback. thanks in advance

bowbahdoe commented:
Sorry im taking so long, as usual my general feedback right now is
* Better than last time
* I don't like linking out to low quality/unreliable websites like geeksforgeeks or random unafiliated blogs. Like, this link is just kinda...bad https://www.protechtraining.com/blog/post/modern-java-for-the-modern-age-by-nick-maiorano-937
* Your "why java" still misses the mark. You talk about OOP and FP at a point in the track where we don't assume they even know what methods are
* Your use of emojis is still...off. I'm not embargoing them but its tonal whiplash
* The words you bold don't exactly make sense to bold. Hard to explain exactly why

bowbahdoe commented:
bumping to keep alive.

firasrg commented:
Hello @bowbahdoe ! thanks for keeping alive! 

I'm sorry I was busy in priorities and holidays. I hope you're doing good; what's new with this PR ?

illuminator3 commented on src/why_java.md:
```diff
@@ -0,0 +1,58 @@
+
+While functional programming **is not** unique to Java, its inclusion in the language provides Java developers with the flexibility to leverage functional programming concepts alongside object-oriented programming!
+
+> As an example, compared to [JavaScript](https://developer.mozilla.org/fr/docs/Web/JavaScript), which is considered a vert popular language, empowered with functional programming, Java is still stronger than JavaScript, as JavaScript [still lacks some object-oriented concepts](https://stackoverflow.com/a/108773). 
```
typo? "vert"

illuminator3 commented on src/why_java.md:
```diff
@@ -0,0 +1,58 @@
+
+The language has evolved over time, incorporating new features and improvements while maintaining backward compatibility with modern business needs. Java's maturity is evident in the stability of its frameworks, patterns, toolset, and practices.
+
+ It has an extensive documentation and resources. You can find comprehensive documentation, tutorials, and examples to help you learn it easily. The vast Java community also provides support through forums, online chat communities, and **open-source** projects.
```
remove the "an"

bowbahdoe commented:
@firasrg Nothing right now. I've been distracted with other things. I'm coming back into book stuff because the playground work is kicking off.
